Lib/importlib/_bootstrap.py

```diff
@@ -512,17 +512,18 @@ def _init_module_attrs(spec, module, *, override=False):
                 loader = NamespaceLoader.__new__(NamespaceLoader)
                 loader._path = spec.submodule_search_locations
                 spec.loader = loader
-                # While the docs say that module.__file__ is not set for
-                # built-in modules, and the code below will avoid setting it if
-                # spec.has_location is false, this is incorrect for namespace
-                # packages.  Namespace packages have no location, but their
-                # __spec__.origin is None, and thus their module.__file__
-                # should also be None for consistency.  While a bit of a hack,
-                # this is the best place to ensure this consistency.
-                #
-                # See # https://docs.python.org/3/library/importlib.html#importlib.abc.Loader.load_module
-                # and bpo-32305
-                module.__file__ = None
+        if _bootstrap_external and isinstance(loader, _bootstrap_external.NamespaceLoader):
```

Comment:
Rescoping this change means it no longer falls under "A backward compatibility hack." Is that intentional?

Reply:
Is the `if loader is None` block above still needed now that `spec.loader` is set in `_bootstrap_external`?

Reply:
I did do a deep dive into this and was meaning to write a more in-depth explanation. If I understand correctly, the backwards compatibility hack is instantiating the loader, not setting `__file__`. This function is meant to fill these missing attributes, but up until now it always expected to be the one instantiating the loader on namespace packages, which now changed with this PR. This should also trip up on 3rd party finders that set the loader for namespace packages, but so far no one has reported it yet.